core/src/main/java/hudson/util/RobustReflectionConverter.java

```diff
@@ -271,8 +271,71 @@ public Object unmarshal(final HierarchicalStreamReader reader, final Unmarshalli
         return serializationMethodInvoker.callReadResolve(result);
     }
 
+    private static final class FieldExpectation {
+        private final Class definingClass;
+        private final String name;
+
+        public FieldExpectation(Class definingClass, String name) {
+            this.definingClass = definingClass;
+            this.name = name;
+        }
+
+        public Class getDefiningClass() {
+            return definingClass;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            FieldExpectation that = (FieldExpectation) o;
+
+            if (definingClass != null ? !definingClass.equals(that.definingClass) : that.definingClass != null) {
+                return false;
+            }
+            return name.equals(that.name);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = definingClass != null ? definingClass.hashCode() : 0;
+            result = 31 * result + name.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "FieldExpectation{" +
+                    "definingClass=" + definingClass +
+                    ", name='" + name + '\'' +
+                    '}';
+        }
+
+
+    }
+
     public Object doUnmarshal(final Object result, final HierarchicalStreamReader reader, final UnmarshallingContext context) {
         final SeenFields seenFields = new SeenFields();
+        final boolean existingObject = context.currentObject() != null;
+        final Map<FieldExpectation, Object> expectedFields = existingObject ? new HashMap<FieldExpectation, Object>() : null;
+        final Object cleanInstance = existingObject ? reflectionProvider.newInstance(result.getClass()) : null;
```

Comment:
Yeah, a pile of NPEs are specifically coming from https://github.com/stephenc/jenkins/blob/c62f4f753d16210b67ec381b9e5d1d60145594d8/core/src/main/java/jenkins/model/Jenkins.java#L3049, and then if that's bypassed, https://github.com/stephenc/jenkins/blob/c62f4f753d16210b67ec381b9e5d1d60145594d8/core/src/main/java/jenkins/model/Jenkins.java#L3095, and so on. The core problem is that the fields are just not getting initialized, and so boom.